src/ucs/arch/generic/cpu.c

```diff
@@ -51,54 +51,62 @@ ucs_status_t ucs_cpu_get_cache_size(ucs_cpu_cache_type_t type, size_t *value)
+        if (file_size < 0) {
+            return UCS_ERR_UNSUPPORTED; /* no more files */
+        }
+
+        if (strcasecmp(ucs_strtrim(type_str), ucs_cpu_cache[type].type)) {
+            continue;
+        }
+
+        status = ucs_read_file_number(&level, 1, UCS_CPU_CACHE_FILE_FMT,
+                                      cpu, i, UCS_CPU_CACHE_LEVEL);
+        if (status != UCS_OK) {
+            return UCS_ERR_UNSUPPORTED; /* no more files */
+        }
+
+        if (level != ucs_cpu_cache[type].level) {
+            continue;
         }
 
-        cpu = ucs_get_first_cpu();
-
-        for (i = 0;; i++) {
-            file_size = ucs_read_file_str(type_str, sizeof(type_str), 1, UCS_CPU_CACHE_FILE_FMT,
-                                          cpu, i, UCS_CPU_CACHE_TYPE);
-            if (file_size < 0) {
-                ucs_cpu_cache[type].status = UCS_ERR_UNSUPPORTED; /* no more files */
-                break;
-            }
-
-            if (strcasecmp(ucs_strtrim(type_str), ucs_cpu_cache[type].type)) {
-                continue;
-            }
-
-            status = ucs_read_file_number(&level, 1, UCS_CPU_CACHE_FILE_FMT,
-                                          cpu, i, UCS_CPU_CACHE_LEVEL);
-            if (status != UCS_OK) {
-                ucs_cpu_cache[type].status = UCS_ERR_UNSUPPORTED; /* no more files */
-                break;
-            }
-
-            if (level != ucs_cpu_cache[type].level) {
-                continue;
-            }
-
-            /* ok, we found valid directory, let's try to read cache size */
-            file_size = ucs_read_file_str(size_str, sizeof(size_str), 1, UCS_CPU_CACHE_FILE_FMT,
-                                          cpu, i, UCS_CPU_CACHE_SIZE);
-            if (file_size < 0) {
-                ucs_cpu_cache[type].status = UCS_ERR_UNSUPPORTED; /* no more files */
-                break;
-            }
-
-            ucs_cpu_cache[type].status = ucs_str_to_memunits(ucs_strtrim(size_str), &ucs_cpu_cache[type].cache);
-            break;
+        /* ok, we found valid directory, let's try to read cache size */
+        file_size = ucs_read_file_str(size_str, sizeof(size_str), 1, UCS_CPU_CACHE_FILE_FMT,
+                                      cpu, i, UCS_CPU_CACHE_SIZE);
+        if (file_size < 0) {
+            return UCS_ERR_UNSUPPORTED; /* no more files */
         }
 
-out:
-        ucs_do_nothing();
+        return ucs_str_to_memunits(ucs_strtrim(size_str), value);
+    }
+}
+
+
+ucs_status_t ucs_cpu_get_cache_size(ucs_cpu_cache_type_t type, size_t *value)
+{
+    ucs_cpu_cache_type_t i;
+
+    if (type >= UCS_CPU_CACHE_LAST) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    UCS_INIT_ONCE(&ucs_cache_read_once) {
+        UCS_STATIC_ASSERT(ucs_countof(ucs_cpu_cache) == UCS_CPU_CACHE_LAST);
+        for (i = UCS_CPU_CACHE_L1d; i < UCS_CPU_CACHE_LAST; i++) {
+            ucs_cpu_cache[i].status = ucs_cpu_read_cache_size(i, &ucs_cpu_cache[i].cache_size);
```

Comment:
why do we need loop inside a loop?
just read all cache sizes at once..

Reply:
the problem is in info source: we are getting info from CPUID & from FS, in general case these ways may be inconsistent (for example fs/kernel info may be incomplete) and it may be problematic to merge info. current implementation reads info which is supported by UCX